[Ready for Review] Restructure to standalone sections (#639)
alan-turing-institute/the-turing-way: book/content/reproducible_environments/06/containers.md

Volume related commands:

- List volumes: `sudo docker volume ls`
- Delete a volume: `sudo docker volume rm volume_name`
- Delete all unattached volumes: `sudo docker volume prune`
- If, when deleting a container a `-v` is included after `rm` in `sudo docker rm container_ID` any volumes associated with the container will also be deleted.

### Singularity

> Prerequisites: At present, Singularity only runs on linux systems (for example Ubuntu). If you use, macOS, [Singularity Desktop for macOS](https://www.sylabs.io/singularity-desktop-macos/) is in "Alpha Preview" stage.

A major drawback of Docker for reproducible research is that it is not intended as a user-space application but as a tool for server administrators. As such it requires root access to operate. There is, however, no reason why the execution of an analysis should require root access for the user. This is especially important when computations are conducted on shared resource like HPC systems where users will never have root access.

The [singularity](https://www.sylabs.io/) container software was introduced to address exactly this issue. Singularity was created with HPC sytems and reproducible research in mind

> (see [this](https://www.youtube.com/watch?v=DA87Ba2dpNM) video]. It does not require root access to run (only to build container _images_!) and thus enables HPC users to locally build container images before running analyses, e.g., on a high-performance cluster. As an added benefit, this makes it possible to use almost any software on an

Reviewer: ```suggestion
(see [this](https://www.youtube.com/watch?v=DA87Ba2dpNM) video). It does not require root access to run (only to build container _images_!) and thus enables HPC users to locally build container images before running analyses, e.g., on a high-performance cluster. As an added benefit, this makes it possible to use almost any software on an
```

HPC system without having to bother admin staff with installing it. In recognition of the fact that Docker is _the_ most well known containerization
approach, singularity aims at maintaining compatibility with docker containers as much as possible, i.e., singularity can be used to run normal docker containers
(without requiring root access!).

Singularity can be used to run Docker images or extend them by building new images based on docker containers as base layer. For instance, we could use singularity to spin up a vanilla ubuntu container and getting a shell in it using the ubuntu docker image via

```
singularity shell docker://ubuntu
```

(type `exit` to leave the interactive shell again).

> Just as docker images are build using `Dockerfile` files, singularity containers

Reviewer: ```suggestion
Just as docker images are built using `Dockerfile` files, singularity containers
```

> are build from singularity definition files.

Reviewer: ```suggestion
are built from singularity definition files.
```

The process and syntax is similar to docker files but there are subtle
differences.
As a minimal working example, we can build a 'lolcow' container based on the
official ubuntu docker container image.
Put the following in a `lolcow.def` file

> (based on the [Singularity documentation](https://www.sylabs.io/guides/3.2/user-guide/build_a_container.html))

Reviewer: ```suggestion
(based on the [Singularity documentation](https://www.sylabs.io/guides/3.2/user-guide/build_a_container.html)):
```

```
Bootstrap: docker
From: ubuntu

%post
    apt-get -y update
    apt-get -y install fortune cowsay lolcat

%environment
    export LC_ALL=C
    export PATH=/usr/games:$PATH

%runscript
    fortune | cowsay | lolcat
```

This 'recipe' uses a docker image as basis (here: ubuntu) installs a few
apt packages, modifies a few environment variables, and specifies the
runscript (which is executed using the `singularity run` command).
Details on the singularity definition file format can be found in the official
[documentation](https://www.sylabs.io/docs/).

> A container image can then be build (requiring root!) via

Reviewer: ```suggestion
A container image can then be built (requiring root!) via
```

```
sudo singularity build lolcow.simg lolcow.def
```

This will pull the ubuntu image from dockerhub, run the steps of the recipe in
the definition file and produce a single output image file (`lolcow.simg`).
Finally the runscript is executed as

```
singularity run lolcow.simg
```

Ideally, you should see a nice ASCII cow and a few words of wisdom, as in

```
___________________________________
/ You will be called upon to help a \
\ friend in trouble.                /
-----------------------------------
       \   ^__^
        \  (oo)\_______
           (__)\       )\/\
               ||----w |
               ||     ||
```

> Being HPC compatible, singularity containers are also supported by wide range

Reviewer: ```suggestion
Being HPC compatible, singularity containers are also supported by a wide range
```

of workflow management tools.

> E.g. both [snakemake](https://snakemake.readthedocs.io/en/stable/) and

Reviewer: ```suggestion
For example, both [snakemake](https://snakemake.readthedocs.io/en/stable/) and
```

[nextflow](https://www.nextflow.io/docs/latest/singularity.html) support
job-specific singularity containers.
This makes singularity containers uniquely suited for parallelizing workflows
on HPC systems using the
widely used [slurm](https://slurm.schedmd.com/documentation.html) workload manager.
Using singularity containers and snakemake/nextflow is therefore a way of
scaling reproducibility to massive scale and - as an added benefit -

> brining workflows from a desktop machine to an HPC system no longer requires

Reviewer: ```suggestion
bringing workflows from a desktop machine to an HPC system no longer requires
```
